src/primitives.rs

```diff
@@ -1,10 +1,167 @@
 //! Contains comparison primitives used to build up the rest of the library.
 
+use core::{
+    cmp::Ordering,
+    convert::{From, TryFrom},
+};
+
+// TODO: turn primitives.rs into
+// primitives/
+//   mod.rs
+//   types.rs for basic types
+//   traits.rs? or maybe per trait
+
+/// Represents a matching streak of characters between two strings.
+///
+/// See [find_longest_match] for details.
+#[derive(PartialEq, Eq, Copy, Clone, Debug)]
+pub struct MatchingStreak {
+    /// The index into the first (typically shorter) string where the streak
+    /// begins.
+    pub idx1: usize,
+    /// The index into the second (typically longer) string where the streak
+    /// begins.
+    pub idx2: usize,
+    /// The size of the matching character streak.
+    pub size: usize,
+}
+
+/// Represents an integer score of 0-100, inclusive.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Clone, Copy)]
+pub struct Score {
+    score: u8,
+}
+
+impl Score {
+    const MAX: u8 = 100;
+    pub fn new(score: u8) -> Option<Self> {
+        if score <= Self::MAX {
+            Some(Score { score })
+        } else {
+            None
+        }
+    }
+    pub fn score(&self) -> u8 {
+        self.score
+    }
+}
+
+impl TryFrom<u8> for Score {
+    // TODO: pick a better 'error' type...
+    type Error = String;
+    fn try_from(score: u8) -> Result<Self, Self::Error> {
+        if score <= Self::MAX {
+            Ok(Score { score })
+        } else {
+            Err(format!("Score must be [0, 100]. Given: {}", score))
+        }
+    }
+}
+
+impl From<Score> for u8 {
+    fn from(score: Score) -> u8 {
+        score.score
+    }
+}
+
+impl PartialEq<u8> for Score {
+    fn eq(&self, other: &u8) -> bool {
+        self.score.eq(other)
+    }
+}
+
+impl PartialOrd<u8> for Score {
+    fn partial_cmp(&self, other: &u8) -> Option<Ordering> {
+        self.score.partial_cmp(other)
+    }
+}
+
+/// Represents a match of some comparison query, with the item and the score of
+/// the match.
+#[derive(PartialEq, Eq, Debug)]
+pub struct Match<T> {
+    pub item: T,
+    pub score: Score,
+}
+
+/// A function of the form `Fn(A) -> B` where `A` is the type of the underlying
+/// object being considered and `B` is the type to be used for scoring the `A`.
+///
+/// For example, this might be used to only compare against the first item of a
+/// string or list. Conceptually, `|a| a.get(0)`.
+pub trait Processor<A, B> {
+    fn process(&self, item: A) -> B;
+}
+
+impl<A, B, F> Processor<A, B> for F
+where
+    F: Fn(A) -> B,
+{
+    fn process(&self, item: A) -> B {
+        self(item)
+    }
+}
+
+/// A function of the form `Fn(A, B) -> Score` used to score choices against an
+/// baseline query.
+// TODO: link wratio
+/// For example, wratio might be used to compare against strings.
+pub trait Scorer<A, B> {
```

Comment:
Nitpick: `Scorer<Q, C>`?

Reply:
except `wratio` doesn't even fit this trait. With the talk in #6, it might be best to do away with traits here and just use `S: Fn(A, B) -> Score` in the relevant signatures in `process.rs`. Then we can maybe collect `simple_ratio`, `best_partial_ratio`, and whatever else into their own `scorer.rs` module.

part of the issue is that the extract methods are unnecessarily generic.

Reply:
>  it might be best to do away with traits here and just use S: Fn(A, B) -> Score in the relevant signatures

👍 The less complexity and code, the better :) 

> part of the issue is that the extract methods are unnecessarily generic.

What would you change about their signatures?

Reply:
I've thought about this some more and think this should get the same treatment as `trait Normalizer` and `trait Segmenter`. The stuff in `fuzzywuzzy_compatible::process` should be able to be adapted to the new trait.

```rust
// Basically takes the output of segmentation on two items...
trait Scorer<T>{
  fn score<T: Eq>(&self, a: &[T], b: &[T]) -> Score;
}

// Then lets implement the most basic `Scorer`, `simple_ratio`...

// ... which happens to need it's a separate trait for reasons ...
trait Aligner {
  fn align<T: Eq>(&self, a: &[T], b: &[T]) ->Vec<(usize, usize, usize)>;
}
pub struct SimpleRatioScorer<A: Aligner>(A);

impl <T: Eq, A: Aligner> Scorer<T> for SimpleRatioScorer<A> {
  fn score<T: Eq>(&self, a: &[T], b: &[T]) -> Score;
    let matches: usize = self.0.align(a, b).iter().map(|&(_, _, s)| s).sum();
    let sumlength = (a.len() + b.len()) as f32;
    if sumlength.is_normal() {
      (100.0 * (2.0 * matches as f32) / sumlength).round().try_into().unwrap()
    } else {
      100.try_into().unwrap()
    }
  }
}

pub struct BasicMatchingBlocks;

impl Aligner for BasicMatchingBlocks {
  fn align<T: Eq>(&self, a: &[T], b: &[T]) ->Vec<(usize, usize, usize)> {
    get_matching_blocks(a, b)
  }
}

pub struct SmithWaterman;

impl Aligner for SmithWaterman{
  fn align<T: Eq>(&self, a: &[T], b: &[T]) ->Vec<(usize, usize, usize)> {
    return magic();
  }
}

pub struct BestPartialScorer<A: Aligner, S: Scorer>(A, S);

impl <T: Eq, A: Aligner, S: Scorer> Scorer<T> for BestPartialScorer<A, S> {
  fn score<T: Eq>(&self, a: &[T], b: &[T]) -> Score;
    let blocks = self.0.align(a, b);
    let mut max = 0.try_into().unwrap();
    for (i, j, _) in blocks {
      let long_start = if j > i { j - i } else { 0 };
      let long_end = std::cmp::min(long_start + shorter.len(), longer.len());
      let long_substr = &longer[long_start..long_end];
      let r = self.1.score(&shorter, long_substr).try_into().unwrap();
      if r > 99 {
        return 100.try_into().unwrap();
      } else if r > max {
        max = r;
      }
    }
    max
  }
}
```

Reply:
also means that token_set_ratio and such can implement the trait and have similar building blocks.